frontend/src/utils/emails/create-email-link.ts

```diff
@@ -0,0 +1,41 @@
+import parse from './template-parser'
+
+function assertEqual(expect: any, expression: any) {
+  if (expect !== expression) {
+    throw new Error(`AssertionFailure: expected ${expect}, got ${expression}`)
+  }
+}
+function constructMailtoUrl(
+  receivers: string[],
+  subject: string,
+  body: string
+): string {
+  let url = 'mailto:'
+  url += receivers.join(',')
+  url += '?subject=' + subject
+  url += '&body=' + encodeURI(body)
+  return url
+}
+
+type ValidTemplates = 'matched' | 'no-match-1' | 'no-match-2'
+export default async function getMailUrl(
+  receivers: string[],
+  template: ValidTemplates,
+  data: string[]
+) {
+  if (template === 'matched') {
+    assertEqual(0, data.length)
+    let { subject, body } = await parse('matching.template', data)
+    return constructMailtoUrl(receivers, subject, body)
+  } else if (template === 'no-match-1') {
+    assertEqual(2, data.length) // [student name, class]
+    let { subject, body } = await parse('no-match-1.template', data)
+    return constructMailtoUrl(receivers, subject, body)
+  } else if (template === 'no-match-2') {
+    assertEqual(1, data.length) // [student name]
+    let { subject, body } = await parse('no-match-2.template', data)
+    return constructMailtoUrl(receivers, subject, body)
+  }
```

Comment:
Maybe a switch statement could work better here, since `type ValidTemplates = 'matched' | 'no-match-1' | 'no-match-2'`. Or at least changing the last `else if` to an `else` so this function returns `string` instead of `string | undefined`.

Reply:
I didn't use an `else` specifically because I didn't want to expose incorrect or under-development mailto links for prod. Maybe an AssertionError would be better suited here so we keep a contract, and then probably show an error toast on the frontend.

Reply:
All use of this function would be on the frontend though, right? So I don't think there's really an issue with exposing mailto links because all inputs to this would be controlled directly as part of the application (not user input). I don't think showing an error toast would be useful because its only purpose would be to show there's an internal development error on the frontend. Aka just don't call the function for templates that are not finished yet